feat(wip): ilp packet lifecycle blog post (#111)
interledger/interledger.org-developers: src/content/blog/2025-01-31-ilp-packet-lifecycle.mdx

---

> layout: ../../layouts/BlogLayout.astro

Reviewer: ```suggestion
```

title: "The Lifecycle of an Interledger Packet"
description: 'A look under the hood of how Rafiki orchestrates an Interledger payment.'
date: 2025-01-31
slug: ilp-packet-lifecycle

> authors: [Nathan Lie]

Reviewer: ```suggestion
authors:
  - Nathan Lie
```

> author_urls: [https://www.linkedin.com/in/nathan-lie-138a73121]

Reviewer: ```suggestion
author_urls:
  - https://www.linkedin.com/in/nathan-lie-138a73121
```

tags:
  - Interledger
---

> import LargeImg from '/src/components/pages/LargeImg.astro'

Reviewer: ```suggestion
import LargeImg from "/src/components/blog/LargeImg.astro";
```

## Introduction
The Interledger Summit just wrapped up recently, along with its celebrations of & discussions on open standards and financial inclusion.

> This year, the humble Interledger Protocol (ILP for short) packet took center stage as Stefan Thomas [gave a keynote](https://www.youtube.com/watch?v=Djw6wMXFv1Q&t=3240s) on how the Interledger Protocol became more fast & reliable by introducing packetization to the protocol.

Reviewer 2: It may be interpreted that packetization was introduced to the protocol recently. Perhaps it can phrased to convey that packetization was a characteristic of the protocol at its inception.

With packetization breaking payments up into small amounts, sending a payment no longer had a single failure point and both parties in a payment would be able to manage risk more effectively.

> Most integrators of the Interledger protocol are leveraging the ILP packet through [Rafiki](https://github.com/interledger/rafiki), as it provides an efficient way for those integrators to participate in an ILP network.

Reviewer: Is there a reason why you link Rafiki to GitHub instead of the docs?

A prospective debitor will use this standard to set payment terms by first creating an _incoming payment_ on the Rafiki instance hosting the receiving wallet address, then acquiring a _quote_ from the Rafiki instance hosting the sending wallet address.
Finally, it creates an _outgoing payment_ against the sending wallet address by acquiring consent from its owner to charge them with particular terms such as its amount and its currency.
Once that is complete, these Rafiki instances can begin to fulfill the terms of this payment using Interledger.

On a more technical point, ILP packets may be involved before an outgoing payment is even created. During the acqusition of a _quote_ from the sending Rafiki instance, it will send a ILP packet accross the network to the destination. This packet will need to traverse each hop on the network as each node's fees and rates won't be known to the sender without doing so.

_For a more on Open Payments, consider reading the [Simple Open Payments Guide](https://interledger.org/developers/blog/simple-open-payments-guide/) on this blog._

### Preparing the Packet

> Now that the sending Rafiki instance has an outgoing payment created against one of its wallet addresses, it can pass the terms of that payment into an Interledger payment. First, it uses the [Simple Payment Setup Protocol](https://github.com/interledger/rfcs/blob/main/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md) to acquire an ILP address for the receiver, and a shared secret that is used to generate a [_condition_](https://github.com/interledger/rfcs/blob/main/0029-stream/0029-stream.md).

Reviewer: ```suggestion
Now that the sending Rafiki instance has an outgoing payment created against one of its wallet addresses, it can pass the terms of that payment into an Interledger payment. First, it uses the [Simple Payment Setup Protocol](/developers/rfcs/simple-payment-setup-protocol/) (SPSP) to acquire an ILP address for the receiver, and a shared secret that is used to generate a [_condition_](/developers/rfcs/stream-protocol/#6-condition-and-fulfillment-generation).
```

> Now that the sending Rafiki instance has an outgoing payment created against one of its wallet addresses, it can pass the terms of that payment into an Interledger payment. First, it uses the [Simple Payment Setup Protocol](https://github.com/interledger/rfcs/blob/main/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md) to acquire an ILP address for the receiver, and a shared secret that is used to generate a [_condition_](https://github.com/interledger/rfcs/blob/main/0029-stream/0029-stream.md).

Reviewer 3: I don't think SPSP is involved during the flow, the incoming payment gets the generated STREAM shared secret (and ILP destination address) from the STREAM "server" on the receiving Rafiki

A condition is a hash that can only be generated with a secret that is available only to the receiver and the sender.

> Using the receiving ILP address, the condition, the receiver, and send amount described in the outgoing payment, Rafiki creates a connection over ILP to the receiver using [STREAM](https://github.com/interledger/rfcs/blob/main/0029-stream/0029-stream.md).

Reviewer: ```suggestion
Using the receiving ILP address, the condition, the receiver, and send amount described in the outgoing payment, Rafiki creates a connection over ILP to the receiver using [STREAM](/developers/rfcs/stream-protocol/).
```

STREAM is a transport-layer protocol for Interledger that is responsible for breaking a payment down into the aforementioned packets and sends them _en masse_ until the payment is fulfilled.

For each packet, STREAM handles them using a two-phase execution process that is similar to the request-response process that HTTP packets sent over the internet use.

> <LargeImg

Reviewer: `![Two-Phase Transfer Diagram](/developers/img/blog/2025-01-31/2p-transfer.png)`

  src="/developers/img/blog/2025-01-31/2p-transfer.png"
  alt="Two-Phase Transfer Diagram"
/>

> In the first phase, the sender creates packets known as ["prepare" packets](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-prepare), which each contain some fraction of the total send amount, the condition generated using SPSP, the final destination of the packet in the network, and an expiration time for the packet.

Reviewer: ```suggestion
In the first phase, the sender creates packets known as ["prepare" packets](/developers/rfcs/interledger-protocol/#ilp-prepare), which each contain some fraction of the total send amount, the condition generated using SPSP, the final destination of the packet in the network, and an expiration time for the packet.
```

> In the first phase, the sender creates packets known as ["prepare" packets](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-prepare), which each contain some fraction of the total send amount, the condition generated using SPSP, the final destination of the packet in the network, and an expiration time for the packet.

Reviewer 3: ```suggestion
In the first phase, the sender creates packets known as ["prepare" packets](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-prepare), which each contain some fraction of the total send amount, the condition generated using SPSP, the final destination of the packet in the network, and an expiration time for the packet.
```
do we want to mention that the fraction is something configured by the peers?

Author: I think the overall story remains coherent without including this detail. I can add a further reading link to https://interledger.org/developers/rfcs/peering-clearing-settling/ at the end, perhaps.

> All of these packets are sent across the network to the receiver, and must be fulfilled or rejected by the receiver in the second phase in order to be considered as having reached the receiver.

Reviewer: Maybe we can simplify this with, "... in order to be considered successful."?

> <LargeImg

Reviewer: `![ILP Prepare Packet Specification](/developers/img/blog/2025-01-31/ilp-prepare-spec.png)`

  src="/developers/img/blog/2025-01-31/ilp-prepare-spec.png"
  alt="ILP Prepare Packet Specification"
/>

### Responding with Fulfill or Reject
In the second phase of STREAM's execution process, the receiver will need to handle each "prepare" packet with either a fulfillment or a rejection, so that the sender knows whether or not a given "prepare" packet was successfully sent or not.

> The receiver fulfills a "prepare" packet by sending a ["fulfill" packet](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-fulfill) back to the sender with the preimage from which the "prepare" packet's condition was generated.

Reviewer: ```suggestion
The receiver fulfills a "prepare" packet by sending a ["fulfill" packet](/developers/rfcs/interledger-protocol/#ilp-fulfill) back to the sender with the preimage from which the "prepare" packet's condition was generated.
```

> <LargeImg

Reviewer: `![ILP Fulfill Packet Specification](/developers/img/blog/2025-01-31/ilp-fulfill-spec.png)`

  src="/developers/img/blog/2025-01-31/ilp-fulfill-spec.png"
  alt="ILP Fulfill Packet Specification"
/>

Since only the receiver can provide the preimage needed to fulfill a "prepare" packet's condition, no node in the network in between it and the sender can intercept the packet and send back a fulfilled packet without it reaching the receiver first.
As soon as the sender receives enough fulfillments from the receiver to complete the payment, it stops sending packets over STREAM and closes the connection.

> If the received "prepare" packet expired before it reached the receiver, the amount received already from prior packets exceeds the total, or some other error occurs, the sender will instead respond with a ["reject" packet](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-reject) instead.

Reviewer: ```suggestion
If the received "prepare" packet expired before it reached the receiver, the amount received already from prior packets exceeds the total, or some other error occurs, the sender will instead respond with a ["reject" packet](/developers/rfcs/interledger-protocol/#ilp-reject) instead.
```

> If the received "prepare" packet expired before it reached the receiver, the amount received already from prior packets exceeds the total, or some other error occurs, the sender will instead respond with a ["reject" packet](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-reject) instead.

Reviewer 3: ```suggestion
If the received "prepare" packet expired before it reached the receiver, the amount received already from prior packets exceeds the total, or some other error occurs, the receiver will instead respond with a ["reject" packet](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md#ilp-reject) instead.
```

This "reject" packet contains an error code that describes what would have caused the rejection.

> <LargeImg

Reviewer: `![ILP Reject Packet Specification](/developers/img/blog/2025-01-31/ilp-reject-spec.png)`

  src="/developers/img/blog/2025-01-31/ilp-reject-spec.png"
  alt="ILP Reject Packet Specification"
/>

> On another technical point, there also exist some types of packets that cannot be fulfilled. For instance, an unfulfillable packet gets sent when an Open Payments quote is created whose purpose is to determine what the total cost of the payment will be by factoring any fees or conversion rates that may be applied.

Reviewer 3: At the end, we should add something along the lines of "...instead of the packet sending actual funds/trying to fulfill a payment"

### Wrapping Things Up
As "fulfill" packets arrive at the sender, it verifies that the preimage sent back in each of them does indeed fulfill the condition sent with the corresponding "prepare" packet.

> Once the STREAM connection finishes sending all of the packets and verifies the corresponding fulfillments from the receiver, it closes. The sending Rafiki instance will then update its outgoing payment as being complete, and then publish a webhook to the receiving Rafiki instance to direct it to mark its corresponding incoming payment as complete as well.

Reviewer 3: > and then publish a webhook to the receiving Rafiki instance

the receiving Rafiki is responsible for sending the "incoming payment compelted" webhook to its ASE

## Conclusion

> The Interledger Protocol had to go through a number of important changes in order to position itself as a web-native way to send money. Packetizing Interledger payments was significant change that brought it in line with the design patterns of the modern web that it seeks to be a part of.

Reviewer: ```suggestion
The Interledger Protocol had to go through a number of important changes in order to position itself as a web-native way to send money. Packetizing Interledger payments was a significant change that brought it in line with the design patterns of the modern web that it seeks to be a part of.
```

> The Interledger Protocol had to go through a number of important changes in order to position itself as a web-native way to send money. Packetizing Interledger payments was significant change that brought it in line with the design patterns of the modern web that it seeks to be a part of.

Reviewer 3: Do we want to specify "web-native"? That implies more browser stuff to me

Reviewer: We could say "future proof" instead, since I believe that was one of the philosophies behind the design?

These packets continue to play this role as a part of Rafiki, as it coordinates those payments with Open Payments and manages the ILP connections used to fulfill them.

If this post has established the connection between Rafiki, Open Payments, and Interledger in a way that piques interest, consider delving futher into these Interledger concepts:

> - [ILP v4](https://github.com/interledger/rfcs/blob/main/0027-interledger-protocol-4/0027-interledger-protocol-4.md)

Reviewer: ```suggestion
- [ILP v4](/developers/rfcs/interledger-protocol/)
```

> - [SPSP](https://github.com/interledger/rfcs/blob/main/0009-simple-payment-setup-protocol/0009-simple-payment-setup-protocol.md)

Reviewer: ```suggestion
- [SPSP](/developers/rfcs/simple-payment-setup-protocol/)
```

> - [STREAM](https://github.com/interledger/rfcs/blob/main/0029-stream/0029-stream.md)

Reviewer: ```suggestion
- [STREAM](/developers/rfcs/stream-protocol/)
```

### Setting the Terms
Before it starts sending all those ILP packets, a higher-level protocol is needed to set the terms of the payment as Interledger doesn't have any mechanism to negotiate payment terms. Rafiki uses the [Open Payments Standard](https://openpayments.dev/introduction/overview/) to achieve this.

A prospective debitor will use this standard to set payment terms by first creating an _incoming payment_ on the Rafiki instance hosting the receiving wallet address, during which the receiving Rafiki instance will associate an ILP address with the incoming payment and generate a shared secret for it.
Then it may acquire a _quote_ from the Rafiki instance hosting the sending wallet address, which becomes associated with the incoming payment.

Finally, it creates an _outgoing payment_ against the sending wallet address by acquiring consent from its owner to charge them with particular terms such as its amount and its currency.
The prior incoming payment (or the quote, if created), is provided during this process and becomes associated with the outgoing payment.
Once that is complete, these Rafiki instances can begin to fulfill the terms of this payment using Interledger.

> On a more technical point, ILP packets may be involved before an outgoing payment is even created. During the acqusition of a _quote_ from the sending Rafiki instance, it will send a ILP packet accross the network to the destination. This packet will need to traverse each hop on the network as each node's fees and rates won't be known to the sender without doing so.

Author: Thanks for catching these!

I will contest though, that the usage of _an_ is correct here, [since it precedes a vowel](https://www.grammarly.com/blog/grammar/indefinite-articles-a-and-an/#:~:text=The%20two%20indefinite%20articles%20in,starting%20with%20a%20vowel%20sound.).

> The two indefinite articles in English are a and an. The indefinite article an is used to make pronunciation easier when reading a text aloud. The general rule is to use a when the indefinite article precedes a word beginning with a consonant sound and an when it precedes a word starting with a vowel sound.

Using the receiving ILP address, the condition, the receiver, and send amount described in the outgoing payment, Rafiki creates a connection over ILP to the receiver using [STREAM](/developers/rfcs/stream-protocol/).
STREAM is a transport-layer protocol for Interledger that is responsible for breaking a payment down into the aforementioned packets and sends them _en masse_ until the payment is fulfilled.

For each packet, STREAM handles them using a two-phase execution process that is similar to the request-response process that HTTP packets sent over the internet use.

![Two-Phase Transfer Diagram](/developers/img/blog/2025-01-31/2p-transfer.png)

In the first phase, the sender creates packets known as [Prepare packets](/developers/rfcs/interledger-protocol/), which each contain some fraction of the total send amount configured by each peer along the ILP network path, the condition generated using SPSP, the final destination of the packet in the network, and an expiration time for the packet.
All of these packets are sent across the network to the receiver, and must be fulfilled or rejected by the receiver in the second phase in order to be considered successfully sent.

> ![ILP Prepare Packet Specification](/developers/img/blog/2025-01-31/ilp-prepare-spec.png)

Reviewer: I wonder if this image could use a title to indicate what the image is of. It doesn't explicitly explain that this is the structure of a prepare packet. And the same for reject and fulfill.

Other than that LGTM.

Once the STREAM connection finishes sending all of the packets and verifies the corresponding fulfillments from the receiver, it closes. The sending Rafiki instance will then update its outgoing payment as being complete, and the receiving Rafiki instance publishes a webhook to its account servicing entity directing it to mark its incoming payment as complete as well.

## Conclusion
The Interledger Protocol had to go through a number of important changes in order to position itself as a future-proof way to send money. Packetizing Interledger payments was a significant change that brought it in line with the design patterns of the modern web that it seeks to be a part of.
These packets continue to play this role as a part of Rafiki, as it coordinates those payments with Open Payments and manages the ILP connections used to fulfill them.

If this post has established the connection between Rafiki, Open Payments, and Interledger in a way that piques interest, consider delving futher into these Interledger concepts:

- [ILP v4](/developers/rfcs/interledger-protocol/)
- [STREAM](/developers/rfcs/stream-protocol/)

> - [Connectors & Peering](/developers/rfcs/peering-clearing-settling)

Reviewer 4: As the ending can you please add the footer from https://interledger.org/developers/blog/rafiki-tigerbeetle-integration/

_As we are open source, you can easily check our work on [GitHub](https://github.com/interledger/). If the work mentioned here inspired you, we welcome your contributions. You can join our [community slack](https://communityinviter.com/apps/interledger/interledger-working-groups-slack) or participate in the next [community call](https://docs.google.com/document/d/1T_Q_eRZtL8GluJR9rVADd7fvkB8mPluMk-d96LM4vlg/edit?usp=sharing), which takes place each second Wednesday of the month.

If you want to stay updated with all open opportunities and news from the Interledger Foundation, you can subscribe to our [newsletter](https://interledger.org/subscribe)._

Author: I think the footer is added automatically. I see it on the dev site when I launch it locally, and the Tigerbeetle post doesn't seem to add anything extra to include the footer [in the source](https://github.com/interledger/interledger.org-developers/blob/main/src/content/blog/2024-12-17-rafiki-tigerbeetle-integration.mdx?plain=1).

Reviewer: @ioana-chiorean they are all added automatically now. So we shouldn't repeat it.
